SSL_write sometimes results in NullPointerException after repeated requests

I haven't been able to narrow this down to the simplest reproducible scenario and this may in fact be a bug in a different library layer (possibly `google-http-java-client`). However, due to discrepancies in behavior between the default OpenJDK implementation and Conscrypt, I suspect this may be a bug in Conscrypt.

Here's the scenario: attempt to write a local file to a Google Cloud Storage (GCS) bucket repeatedly. Writing a small file 10 times appears to work. However, with a 1 GiB file of random data, this appears to deterministically fail due to `ssl == null` on the third attempt.

Example program that reproduces this behavior:

```java
import com.google.api.gax.paging.Page;
import com.google.cloud.WriteChannel;
import com.google.cloud.storage.Blob;
import com.google.cloud.storage.BlobInfo;
import com.google.cloud.storage.BucketInfo;
import com.google.cloud.storage.Storage;
import com.google.cloud.storage.StorageOptions;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.channels.Channels;
import java.nio.channels.FileChannel;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

public class Main {
  public static void main(String[] args) throws Exception {
    String inputPath = args[0];
    int chunkSize = Integer.getInteger("chunk.size", 100*1024*1024);
    int bufferSize = Integer.getInteger("buffer.size", 1024*16);
    if (bufferSize <= 0) {
      throw new RuntimeException("Invalid buffer size: " + bufferSize);
    }
    Storage storage = StorageOptions.getDefaultInstance().getService();

    String bucketName = "bucketname";
    String objectName = "objectname";

    BlobInfo blobInfo = BlobInfo.newBuilder(BucketInfo.of(bucketName), objectName).build();
    Path path = Paths.get(inputPath);

    for (int i = 0; i < 10; i++) {
      try (WriteChannel output = storage.writer(blobInfo);
          FileChannel input = FileChannel.open(path, StandardOpenOption.READ);
          ) {
        output.setChunkSize(chunkSize);
        ByteBuffer buffer = ByteBuffer.allocateDirect(bufferSize);
        while (input.read(buffer) != -1) {
          buffer.flip();
          output.write(buffer);
          buffer.compact();
        }
        while (buffer.remaining() > 0) {
          int n = output.write(buffer);
          if (n <= 0) {
            throw new RuntimeException("No bytes written");
          }
        }
      }
      System.out.println("Finished iteration " + i);
    }
  }
}
```
Output of a failing attempt (including stack trace):
```
Finished iteration 0
Finished iteration 1
Exception in thread "main" com.google.cloud.storage.StorageException: java.lang.NullPointerException: ssl == null
        at com.google.cloud.storage.StorageException.translateAndThrow(StorageException.java:71)
        at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:55)
        at com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:121)
        at com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:148)
        at com.google.gcsjava.Main.main(Main.java:56)
        Suppressed: com.google.cloud.storage.StorageException: java.lang.NullPointerException: ssl == null
                at com.google.cloud.storage.StorageException.translateAndThrow(StorageException.java:71)
                at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:55)
                at com.google.cloud.BaseWriteChannel.close(BaseWriteChannel.java:160)
                at com.google.gcsjava.Main.main(Main.java:65)
        Caused by: java.lang.NullPointerException: ssl == null
                at org.conscrypt.NativeCrypto.SSL_write(Native Method)
                at org.conscrypt.SslWrapper.write(SslWrapper.java:380)
                at org.conscrypt.ConscryptFileDescriptorSocket$SSLOutputStream.write(ConscryptFileDescriptorSocket.java:617)
                at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
                at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
                at java.io.PrintStream.flush(PrintStream.java:338)
                at sun.net.www.MessageHeader.print(MessageHeader.java:301)
                at sun.net.www.http.HttpClient.writeRequests(HttpClient.java:644)
                at sun.net.www.http.HttpClient.writeRequests(HttpClient.java:655)
                at sun.net.www.protocol.http.HttpURLConnection.writeRequests(HttpURLConnection.java:693)
                at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1326)
                at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1291)
                at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:250)
                at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
                at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
                at com.google.cloud.storage.spi.v1.HttpStorageRpc.write(HttpStorageRpc.java:562)
                at com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:51)
                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
                at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:94)
                at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:54)
                at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:47)
                ... 2 more
Caused by: java.lang.NullPointerException: ssl == null
        at org.conscrypt.NativeCrypto.SSL_write(Native Method)
        at org.conscrypt.SslWrapper.write(SslWrapper.java:380)
        at org.conscrypt.ConscryptFileDescriptorSocket$SSLOutputStream.write(ConscryptFileDescriptorSocket.java:617)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at java.io.PrintStream.flush(PrintStream.java:338)
        at sun.net.www.MessageHeader.print(MessageHeader.java:301)
        at sun.net.www.http.HttpClient.writeRequests(HttpClient.java:644)
        at sun.net.www.http.HttpClient.writeRequests(HttpClient.java:655)
        at sun.net.www.protocol.http.HttpURLConnection.writeRequests(HttpURLConnection.java:693)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1326)
        at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1291)
        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:250)
        at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:77)
        at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:981)
        at com.google.cloud.storage.spi.v1.HttpStorageRpc.write(HttpStorageRpc.java:562)
        at com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:51)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:94)
        at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:54)
        at com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:47)
        ... 3 more

```

The fact that the failure appears deterministic with large files tells me that this probably has something to do with the [finalizer on ConscryptFileDescriptorSocket](https://github.com/google/conscrypt/blob/07f3b1f1ef63cf93d7d5fe7292ed08947d565341/common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java#L985). This seems to be the only place where [`SslWrapper#close()`](https://github.com/google/conscrypt/blob/07f3b1f1ef63cf93d7d5fe7292ed08947d565341/common/src/main/java/org/conscrypt/SslWrapper.java#L499) is called, which in turn is the only place where the NativeCrypto pointer is zeroed out.

This could be due to a bug in client code where a socket/connection is incorrectly reused. However it's also possible that the SSL context is being closed when it shouldn't be.

There are 2 additional points to note:
* When uploading a 1 GiB file multiple times, the second write appears to succeed, but the hash of the GCS contents does _not_ match the expected value. The upload appears to have been clobbered somehow.
* Using the default OpenJDK SSL provider succeeds no matter how many uploads are attempted and no matter the file size.

--- a/common/src/main/java/org/conscrypt/SslWrapper.java
+++ b/common/src/main/java/org/conscrypt/SslWrapper.java
@@ -29,7 +29,6 @@ import java.io.FileDescriptor;
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
 import java.net.SocketException;
-import java.net.SocketTimeoutException;
 import java.security.InvalidKeyException;
 import java.security.PrivateKey;
 import java.security.PublicKey;

--- a/common/src/main/java/org/conscrypt/SslWrapper.java
+++ b/common/src/main/java/org/conscrypt/SslWrapper.java
@@ -28,6 +28,7 @@ import static org.conscrypt.NativeConstants.SSL_VERIFY_PEER;
 import java.io.FileDescriptor;
 import java.io.IOException;
 import java.io.UnsupportedEncodingException;
+import java.net.SocketException;
 import java.net.SocketTimeoutException;
 import java.security.InvalidKeyException;
 import java.security.PrivateKey;
@@ -363,7 +364,10 @@ final class SslWrapper {
 
     // TODO(nathanmittler): Remove once after we switch to the engine socket.
     void doHandshake(FileDescriptor fd, int timeoutMillis)
-            throws CertificateException, SocketTimeoutException, SSLException {
+            throws CertificateException, IOException {
+        if (isClosed() || fd == null || !fd.valid()) {
+            throw new SocketException("Socket is closed");
+        }
         NativeCrypto.SSL_do_handshake(ssl, fd, handshakeCallbacks, timeoutMillis);
     }
 
@@ -374,12 +378,18 @@ final class SslWrapper {
     // TODO(nathanmittler): Remove once after we switch to the engine socket.
     int read(FileDescriptor fd, byte[] buf, int offset, int len, int timeoutMillis)
             throws IOException {
+        if (isClosed() || fd == null || !fd.valid()) {
+            throw new SocketException("Socket is closed");
+        }
         return NativeCrypto.SSL_read(ssl, fd, handshakeCallbacks, buf, offset, len, timeoutMillis);
     }
 
     // TODO(nathanmittler): Remove once after we switch to the engine socket.
     void write(FileDescriptor fd, byte[] buf, int offset, int len, int timeoutMillis)
             throws IOException {
+        if (isClosed() || fd == null || !fd.valid()) {
+            throw new SocketException("Socket is closed");
+        }
         NativeCrypto.SSL_write(ssl, fd, handshakeCallbacks, buf, offset, len, timeoutMillis);
     }
 

--- a/openjdk-integ-tests/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java
+++ b/openjdk-integ-tests/src/test/java/libcore/javax/net/ssl/SSLSocketTest.java
@@ -2457,6 +2457,30 @@ public class SSLSocketTest extends AbstractSSLTest {
         }
     }
 
+    // Under some circumstances, the file descriptor socket may get finalized but still
+    // be reused by the JDK's built-in HTTP connection reuse code.  Ensure that a
+    // SocketException is thrown if that happens.
+    @Test
+    public void test_SSLSocket_finalizeThrowsProperException() throws Exception {
+        TestSSLSocketPair test = TestSSLSocketPair.create().connect();
+        try {
+            if (isConscryptFdSocket(test.client)) {
+                SSLSocket underlying = unwrap(test.client);
+                Method method = underlying.getClass().getDeclaredMethod("finalize");
+                method.setAccessible(true);
+                method.invoke(underlying);
+                try {
+                    test.client.getOutputStream().write(new byte[] { 0x01 });
+                    fail("The socket shouldn't work after being finalized");
+                } catch (SocketException expected) {
+                    // Expected
+                }
+            }
+        } finally {
+            test.close();
+        }
+    }
+
     private static void setWriteTimeout(Object socket, int timeout) {
         Exception ex = null;
         try {

